src/main/java/org/walkmod/javalang/compiler/types/ASMClass.java

```diff
@@ -0,0 +1,73 @@
+package org.walkmod.javalang.compiler.types;
+
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.tree.ClassNode;
+
+import java.io.File;
+
+
+public class ASMClass extends ClassNode {
+
+
+    private String actualName;
+    private Boolean anonymous;
+    private boolean isPrivate = false;
+
+
+    public ASMClass() {
+        super(Opcodes.ASM5);
+    }
+
+    @Override
+    public void visit(int version, int access,
+                      String name, String signature, String superName, String[] interfaces) {
+        super.visit(version, access, name, signature, superName, interfaces);
+        actualName = name;
+        isPrivate = access == Opcodes.ACC_PRIVATE;
+    }
+
+    @Override
+    public void visitInnerClass(String name, String outer, String innerName, int access) {
+        super.visitInnerClass(name, outer, innerName, access);
+        if (name.equals(actualName)) {
+            anonymous = innerName == null;
+        }
+    }
+
+    public boolean isAnonymous() {
+        return anonymous != null && anonymous;
+    }
+
+    public boolean isMemberClass() {
+        return anonymous != null && !anonymous;
+    }
+
+    public String getActualName() {
+        return actualName;
+    }
+
+    public boolean isPrivate() {
+        return isPrivate;
+    }
+
+    public String getPackage() {
+        if (!actualName.contains(File.separator)){
+            return null;
+        }
+        return getActualName().substring(0, getActualName().lastIndexOf("/") + 1)
+                .replaceAll(File.separator, "\\.");
+    }
+
+    public String getSimpleName() {
+        int index = getActualName().lastIndexOf(File.separator);
+        if (index != -1){
+            return actualName;
+        } else {
+            return getActualName().substring(index + 1);
+        }
+    }
+
+    public String getCanonicalName() {
+        return getPackage() + "." + getSimpleName().replaceAll("$", "\\.");
+    }
```

Comment:
Must be "replace"

Reply:
I do not agree. I may have multiple nesting levels, right?

Reply:
We had the same discussion 2 (?) years ago. :-)
"replace(a,b)" replaces *all* strings "a" with string "b".
replaceAll(regular-expression, replacement) replaces all occurences of the regular expression. And the regexp "$" means "end of line".

Reply:
Oops 😄